File: src/core/radar_system.py
#!/usr/bin/env python3
"""
High-Performance Multi-Radar System - Core Module
Handles live streaming and bin replay with 50-60Hz+ performance
"""
import threading
import queue
import time
import csv
import sys
import os
from collections import deque
from dataclasses import dataclass
from typing import Optional, Dict, List, Union

@dataclass
class RadarFrame:
    radar_id: int
    timestamp: float
    raw_data: bytes
    parsed_data: Optional[Dict] = None

class RingBuffer:
    def __init__(self, size: int = 30000):
        self.buffer = deque(maxlen=size)
        self.lock = threading.Lock()
        self.dropped_frames = 0
    
    def put(self, item):
        with self.lock:
            if len(self.buffer) == self.buffer.maxlen:
                self.dropped_frames += 1
            self.buffer.append(item)
    
    def get_all(self):
        with self.lock:
            items = list(self.buffer)
            self.buffer.clear()
            return items

class RadarSystem:
    def __init__(self):
        self.sources = []
        self.buffers = []
        self.processor = None
        self.output_queue = queue.Queue(maxsize=5000)
        self.radar_flags = {}
        self.csv_writer = None
        self.csv_file = None
        self.start_time = None
        
    def add_source(self, source):
        buffer = RingBuffer()
        flag = threading.Event()
        flag.set()
        
        source.setup(buffer, flag)
        self.sources.append(source)
        self.buffers.append(buffer)
        self.radar_flags[source.radar_id] = flag
        
    def start(self, csv_filename: Optional[str] = None, console_output: bool = True):
        if csv_filename:
            self.csv_file = open(csv_filename, 'w', newline='')
            self.csv_writer = csv.writer(self.csv_file)
            self.csv_writer.writerow(['timestamp', 'radar_id', 'frame_number', 'angle', 'distance', 'velocity', 'x', 'y', 'z', 'peak_val', 'range_idx', 'doppler_idx'])
        
        from core.data_processor import DataProcessor
        self.processor = DataProcessor(self.buffers, self.output_queue)
        self.processor.start()
        
        for source in self.sources:
            source.start()
        
        self.start_time = time.time()
        print(f"[SYSTEM] Started with {len(self.sources)} radar sources")
        
        self._main_loop(console_output)
    
    def _main_loop(self, console_output: bool):
        frame_count = 0
        last_stats = time.time()
        
        try:
            while True:
                try:
                    frame = self.output_queue.get(timeout=0.1)
                    frame_count += 1
                    
                    if frame.parsed_data:
                        if console_output:
                            self._display_frame_data(frame.parsed_data)
                        
                        if self.csv_writer:
                            self._log_frame_data(frame)
                            if frame_count % 100 == 0:
                                self.csv_file.flush()
                    
                    if time.time() - last_stats > 3.0:
                        elapsed = time.time() - self.start_time
                        hz = frame_count / elapsed if elapsed > 0 else 0
                        print(f"[PERF] {hz:.1f} Hz ({frame_count} frames)")
                        last_stats = time.time()
                        
                except queue.Empty:
                    continue
                    
        except KeyboardInterrupt:
            print("\n[SYSTEM] Stopping...")
            self.stop()
    
    def stop(self):
        if self.processor:
            self.processor.stop()
        for source in self.sources:
            source.stop()
        if self.csv_file:
            self.csv_file.close()
    
    def _display_frame_data(self, data: Dict):
        """Display frame data based on type"""
        data_type = data.get('data_type', 'detected_object')
        
        if data_type == 'detected_object':
            print(f"R{data['radar_id']}: {data['angle']:6.1f}° {data['distance']:6.1f}cm vel:{data.get('velocity', 0):5.1f} peak:{data.get('peak_val', 0)}")
        elif data_type == 'noise_profile':
            print(f"R{data['radar_id']}: NOISE avg={data['avg_noise_level']:.1f} samples={data['samples']}")
        elif data_type == 'azimuth_heatmap':
            print(f"R{data['radar_id']}: HEATMAP max={data['max_intensity']} avg={data['avg_intensity']:.1f}")
    
    def _log_frame_data(self, frame):
        """Log frame data to CSV based on type"""
        data = frame.parsed_data
        data_type = data.get('data_type', 'detected_object')
        
        if data_type == 'detected_object':
            # Log complete object data
            self.csv_writer.writerow([
                frame.timestamp, frame.radar_id, data['frame_number'],
                data['angle'], data['distance'], data.get('velocity', 0),
                data.get('x', 0), data.get('y', 0), data.get('z', 0),
                data.get('peak_val', 0), data.get('range_idx', 0), data.get('doppler_idx', 0)
            ])
        elif data_type == 'noise_profile':
            self.csv_writer.writerow([
                frame.timestamp, frame.radar_id, 0,
                0, 0, data['avg_noise_level'], 0, 0, 0, 0, 0, 0
            ])
        elif data_type == 'azimuth_heatmap':
            self.csv_writer.writerow([
                frame.timestamp, frame.radar_id, 0,
                0, 0, data['max_intensity'], 0, 0, 0, 0, 0, 0
            ])

File: cpp/src/core/radar_system.cpp
#include "core/radar_system.h"
#include "core/ring_buffer.h"
#include <iostream>
#include <chrono>
#include <iomanip>
#include <cmath>

RadarSystem::RadarSystem() 
    : running_(false), start_time_(0.0), frame_count_(0), ego_vehicle_speed_(0.0) {
}

RadarSystem::~RadarSystem() {
    stop();
}

void RadarSystem::add_source(std::unique_ptr<RadarSource> source) {
    auto buffer = std::make_shared<RingBuffer<RadarFrame>>();
    source->setup(buffer);
    
    sources_.push_back(std::move(source));
    buffers_.push_back(buffer);
    radar_active_flags_.push_back(true);
}

void RadarSystem::start(const std::string& csv_filename, bool console_output) {
    // Setup separate CSV files for detected and tracked objects
    if (!csv_filename.empty()) {
        std::string detected_file = csv_filename.substr(0, csv_filename.find_last_of('.')) + "_detected.csv";
        std::string tracked_file = csv_filename.substr(0, csv_filename.find_last_of('.')) + "_tracked.csv";
        
        detected_csv_file_.open(detected_file);
        tracked_csv_file_.open(tracked_file);
        
        if (detected_csv_file_.is_open()) {
            detected_csv_file_ << "timestamp,radar_id,frame_number,angle,distance,velocity,x,y,z,peak_val,range_idx,doppler_idx\n";
        }
        if (tracked_csv_file_.is_open()) {
            tracked_csv_file_ << "timestamp,radar_id,frame_number,angle,distance,velocity,x,y,z,peak_val,range_idx,doppler_idx,track_id\n";
        }
    }
    
    // Initialize data processor with buffers and output queue
    processor_ = std::make_unique<DataProcessor>(buffers_, output_queue_);
    processor_->start();
    
    running_ = true;
    
    // Start all active sources
    for (size_t i = 0; i < sources_.size(); ++i) {
        if (radar_active_flags_[i]) {
            sources_[i]->start();
        }
    }
    
    auto now = std::chrono::high_resolution_clock::now();
    start_time_ = std::chrono::duration<double>(now.time_since_epoch()).count();
    
    std::cout << "[SYSTEM] Started with " << sources_.size() << " radar sources (optimized C++ implementation)" << std::endl;
    
    main_loop(console_output);
}

void RadarSystem::stop() {
    running_ = false;
    
    if (processor_) {
        processor_->stop();
    }
    
    for (auto& source : sources_) {
        source->stop();
    }
    
    if (detected_csv_file_.is_open()) {
        detected_csv_file_.close();
    }
    if (tracked_csv_file_.is_open()) {
        tracked_csv_file_.close();
    }
}

void RadarSystem::set_ego_vehicle_speed(double speed_mps) {
    ego_vehicle_speed_ = speed_mps;
    std::cout << "[CAN] Ego vehicle speed updated: " << speed_mps << " m/s" << std::endl;
}

void RadarSystem::activate_radar(int radar_id, bool activate) {
    if (radar_id >= 0 && radar_id < static_cast<int>(radar_active_flags_.size())) {
        radar_active_flags_[radar_id] = activate;
        std::cout << "[FIRMWARE] Radar " << radar_id << (activate ? " activated" : " deactivated") << std::endl;
        
        if (running_) {
            if (activate) {
                sources_[radar_id]->start();
            } else {
                sources_[radar_id]->stop();
            }
        }
    }
}

void RadarSystem::main_loop(bool console_output) {
    auto last_stats = std::chrono::high_resolution_clock::now();
    
    while (running_) {
        bool processed_any = false;
        
        // Process data from output queue
        std::lock_guard<std::mutex> lock(queue_mutex_);
        while (!output_queue_.empty()) {
            RadarFrame frame = output_queue_.front();
            output_queue_.pop();
            frame_count_++;
            processed_any = true;
            
            if (!frame.parsed_data.empty()) {
                for (const auto& obj : frame.parsed_data) {
                    if (console_output) {
                        display_frame_data(obj);
                    }
                    
                    // Log to appropriate CSV file
                    if (!is_static_object(obj)) {
                        log_detected_objects(frame);
                        // Simple tracking logic - in production use Kalman filter
                        log_tracked_objects(frame);
                    }
                }
                
                if (frame_count_ % 100 == 0) {
                    if (detected_csv_file_.is_open()) detected_csv_file_.flush();
                    if (tracked_csv_file_.is_open()) tracked_csv_file_.flush();
                }
            }
        }
        
        // Performance statistics
        auto now = std::chrono::high_resolution_clock::now();
        auto stats_elapsed = std::chrono::duration<double>(now - last_stats).count();
        
        if (stats_elapsed > 3.0) {
            auto total_elapsed = std::chrono::duration<double>(now.time_since_epoch()).count() - start_time_;
            double hz = frame_count_ / total_elapsed;
            std::cout << "[PERF] " << std::fixed << std::setprecision(1) << hz 
                     << " Hz (" << frame_count_ << " frames)" << std::endl;
            last_stats = now;
        }
        
        if (!processed_any) {
            std::this_thread::sleep_for(std::chrono::microseconds(100));
        }
    }
}

void RadarSystem::display_frame_data(const DetectedObject& obj) {
    std::cout << "R" << obj.radar_id << ": " 
              << std::fixed << std::setprecision(1)
              << obj.angle << "° " << obj.distance << "cm vel:" 
              << obj.velocity << " peak:" << obj.peak_val << std::endl;
}

void RadarSystem::log_detected_objects(const RadarFrame& frame) {
    if (!detected_csv_file_.is_open()) return;
    
    for (const auto& obj : frame.parsed_data) {
        detected_csv_file_ << frame.timestamp << "," << frame.radar_id << "," 
                          << obj.frame_number << "," << obj.angle << "," << obj.distance << "," 
                          << obj.velocity << "," << obj.x << "," << obj.y << "," << obj.z << "," 
                          << obj.peak_val << "," << obj.range_idx << "," << obj.doppler_idx << "\n";
    }
}

void RadarSystem::log_tracked_objects(const RadarFrame& frame) {
    if (!tracked_csv_file_.is_open()) return;
    
    // Simple tracking - in production use proper tracking algorithm
    static int track_id = 0;
    for (const auto& obj : frame.parsed_data) {
        tracked_csv_file_ << frame.timestamp << "," << frame.radar_id << "," 
                         << obj.frame_number << "," << obj.angle << "," << obj.distance << "," 
                         << obj.velocity << "," << obj.x << "," << obj.y << "," << obj.z << "," 
                         << obj.peak_val << "," << obj.range_idx << "," << obj.doppler_idx << "," 
                         << track_id++ << "\n";
    }
}

bool RadarSystem::is_static_object(const DetectedObject& obj) {
    // Filter static objects using ego vehicle speed
    // Object is static if its relative velocity compensates for ego motion
    double relative_velocity = std::abs(obj.velocity);
    double ego_compensation = ego_vehicle_speed_ * std::cos(obj.angle * M_PI / 180.0);
    
    return std::abs(relative_velocity - ego_compensation) < 0.5; // 0.5 m/s threshold
}